src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReferenceEqualsWrapper.cs

```diff
@@ -0,0 +1,14 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json.Serialization
+{
+    internal struct ReferenceEqualsWrapper : IEquatable<ReferenceEqualsWrapper>
+    {
+        private object _object;
+        public ReferenceEqualsWrapper(object obj) => _object = obj;
+        public override bool Equals(object? obj) => obj is ReferenceEqualsWrapper otherObj && Equals(otherObj);
+        public bool Equals(ReferenceEqualsWrapper obj) => ReferenceEquals(_object, obj._object);
+        public override int GetHashCode() => _object.GetHashCode();
```

Comment:
```suggestion
        public override int GetHashCode() => RuntimeHelpers.GetHashCode(_object);
```

Hashcode equality should always imply equality, so dispatching to whatever hashcode implementation the underlying object encapsulates is breaking that invariant.